neutone_sdk/filters.py

```diff
@@ -0,0 +1,272 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import math
+from typing import List, Optional
+
+"""
+Filters for pre-filtering inputs to models such as RAVE.
+"""
+
+
+class FIRFilter(nn.Module):
+    def __init__(
+        self,
+        cutoffs: List[float],
+        sample_rate: int,
+        filt_size: int = 257,
+        filt_type: str = "lowpass",
+    ):
+        """Streamable FIR filter for pre-filtering of model inputs, etc.
+
+        Args:
+            cutoffs (List[float]): Cutoff frequencies (in Hz). 2 should be given if bandpass/stop
+            sample_rate (int): Sampling rate
+            filt_size (int, optional): Length of the FIR. Defaults to 257.
+            filt_type (str, optional): Type of the filter (low/high/bandpass, bandstop). Defaults to "bandpass".
+        """
+        super().__init__()
+        # register buffer only allowed once
+        self.register_buffer("cache", torch.zeros(2, filt_size - 1))
+        self.register_buffer("ir_windowed", torch.empty(1, 1, filt_size))
+        self.set_parameters(cutoffs, sample_rate, filt_size, filt_type)
+
+    def set_parameters(
+        self,
+        cutoffs: Optional[List[float]] = None,
+        sample_rate: Optional[int] = None,
+        filt_size: Optional[int] = None,
+        filt_type: Optional[str] = None,
+    ):
+        cutoffs = self.cutoffs if cutoffs is None else cutoffs
+        sample_rate = self.sample_rate if sample_rate is None else sample_rate
+        filt_size = self.filt_size if filt_size is None else filt_size
+        filt_type = self.filt_type if filt_type is None else filt_type
+        if len(cutoffs) == 2:
+            if filt_type in ["highpass", "lowpass"]:
+                raise ValueError("only 1 cutoff value supported for this filter type")
```

Comment:
ValueError(f"Only 1 cutoff value supported for filter type {filt_type}"). Same for the one below